Windows - Cannot run Pact

After running the pact release for Windows (v0.0.3) and verifying that the daemon was listening on 6666, the consumer example could not connect to the daemon therefore failing the consumer tests.

`pact-go/dsl/client.go`
net.Dial and net.DialHttp issues with connecting to localhost. Specifying localhost solved the issue in the server name. This may possibly be linked to IPv6.

After modifying `client.go` and the daemon logs the connection we have this issue:
```

2017/02/09 13:50:44 [DEBUG] daemon - starting mock server

2017/02/09 13:50:44 [DEBUG] starting mock service on port: 53215

2017/02/09 13:50:44 [DEBUG] starting service

2017/02/09 13:50:45 [INFO] service: ERROR: "pact-mock-service.rb service" was called with arguments ["--port 53215", "--pact-specification-version 2", "--pact-dir H:\\pact-workshop\\pact-go-consumer\\pacts", "--log H:\\pact-workshop\\pact-go-consumer\\logs/pact.log", "--consumer ui", "--provider ruleservice"]

2017/02/09 13:50:45 [INFO] service: Usage: "pact-mock-service.rb service"

```
It looks like the pact-mock-service.rb doesn't like Windows paths.

1. First i modified `client.go` and changed the Dial and DialHTTP to point to `localhost` so that it could connect to the Daemon.
2. Modified the `consumer.go` file to use DEBUG log level
3. Modified `pact.go` and manually specified log file with:
```go
if p.Server == nil {
		p.LogDir = filepath.FromSlash("C:/Users/Tony/Projects/Go/src/github.com/pact-foundation/pact-go/examples/log/pact.log")
		args := []string{
			fmt.Sprintf("--pact-specification-version %d", p.SpecificationVersion),
			fmt.Sprintf("--pact-dir %s", p.PactDir),
			fmt.Sprintf("--log %s", p.LogDir),
			fmt.Sprintf("--consumer %s", p.Consumer),
			fmt.Sprintf("--provider %s", p.Provider),
		}
		client := &PactClient{Port: p.Port}
		p.pactClient = client

		fmt.Println("ARGUMENTS ARE:", args)
		p.Server = client.StartServer(args)
	}
```

Also printed out the arguments to see what they were:
```
2017/02/13 11:01:27 [DEBUG] pact setup logging
2017/02/13 11:01:27 [DEBUG] pact setup
ARGUMENTS ARE: [--pact-specification-version 2 --pact-dir C:\Users\Tony\Projects\go\src\github.com\pact-foundation\pact-go\examples\pacts --log C:\Users\Tony\Projects\Go\src\github.com\pact-foundation\pact-go\examples\log\pact.log --consumer MyConsumer --provider MyProvider]
2017/02/13 11:01:27 [DEBUG] client: starting a server
2017/02/13 11:01:27 [DEBUG] creating an HTTP client
2017/02/13 11:01:27 [DEBUG] waiting for port 6666 to become available
2017/02/13 11:01:27 [DEBUG] waiting for port 58166 to become available
2017/02/13 11:01:29 [ERROR] Expected server to start < 1s. Timed out waiting for Mock Server to
                        start on port 58166 - are you sure it's running?
2017/02/13 11:01:29 [DEBUG] pact add interaction
2017/02/13 11:01:29 [DEBUG] pact setup logging
2017/02/13 11:01:29 [DEBUG] pact setup
2017/02/13 11:01:29 [DEBUG] pact add interaction
2017/02/13 11:01:29 [DEBUG] pact setup logging
2017/02/13 11:01:29 [DEBUG] pact setup
2017/02/13 11:01:29 [DEBUG] pact verify
2017/02/13 11:01:29 [DEBUG] mock service add interaction
2017/02/13 11:01:31 Error on Verify: Post http://localhost:58166/interactions: dial tcp [::1]:58166: connectex: No connection could be made because the target machine actively refused it.

```
Daemon was run with --logLevel DEBUG:
```
$ ./pact-go_windows_386.exe daemon --logLevel DEBUG
2017/02/13 11:01:21 [DEBUG] setting up a service manager
2017/02/13 11:01:21 [DEBUG] setting up a service manager
2017/02/13 11:01:21 [DEBUG] setting up a service manager
2017/02/13 11:01:21 [DEBUG] setting up a service manager
2017/02/13 11:01:21 [INFO] daemon - starting daemon on port 6666
2017/02/13 11:01:21 [DEBUG] starting service removal monitor
2017/02/13 11:01:21 [DEBUG] starting service creation monitor
2017/02/13 11:01:21 [DEBUG] starting service removal monitor
2017/02/13 11:01:21 [DEBUG] starting service creation monitor
2017/02/13 11:01:21 [DEBUG] starting service creation monitor
2017/02/13 11:01:21 [DEBUG] starting service removal monitor
2017/02/13 11:01:21 method Shutdown has wrong number of ins: 1
2017/02/13 11:01:21 method StartDaemon has wrong number of ins: 2
2017/02/13 11:01:21 [DEBUG] starting service removal monitor
2017/02/13 11:01:21 [DEBUG] starting service creation monitor
2017/02/13 11:01:27 [DEBUG] daemon - starting mock server
2017/02/13 11:01:27 [DEBUG] starting mock service on port: 58166
2017/02/13 11:01:27 [DEBUG] starting service
2017/02/13 11:01:28 [INFO] service: ERROR: "pact-mock-service.rb service" was called with arguments ["--port 58166", "--pact-specification-version 2", "--pact-dir C:\\Users\\Tony\\Projects\\go\\src\\github.com\\pact-foundation\\pact-go\\examples\\pacts", "--log C:\\Users\\Tony\\Projects\\go\\src\\github.com\\pact-foundation\\pact-go\\examples\\log\\pact.log", "--consumer MyConsumer", "--provider MyProvider"]
2017/02/13 11:01:28 [INFO] service: Usage: "pact-mock-service.rb service"
```

--- a/command/daemon.go
+++ b/command/daemon.go
@@ -6,6 +6,8 @@ import (
 )
 
 var port int
+var network string
+var address string
 var daemonCmdInstance *daemon.Daemon
 var daemonCmd = &cobra.Command{
 	Use:   "daemon",
@@ -19,11 +21,13 @@ var daemonCmd = &cobra.Command{
 		verifier := &daemon.VerificationService{}
 		verifier.Setup()
 		daemonCmdInstance = daemon.NewDaemon(mock, verifier)
-		daemonCmdInstance.StartDaemon(port)
+		daemonCmdInstance.StartDaemon(port, network, address)
 	},
 }
 
 func init() {
-	daemonCmd.Flags().IntVarP(&port, "port", "p", 6666, "Local daemon port")
+	daemonCmd.Flags().IntVarP(&port, "port", "p", 6666, "Local daemon port to listen on")
+	daemonCmd.Flags().StringVarP(&network, "network", "n", "", "Local network interface to listen on ('tcp', 'tcp4', 'tcp6')")
+	daemonCmd.Flags().StringVarP(&address, "address", "a", "", "Local network address to listen on (e.g. '', '127.0.0.1', '[::1]' etc.)")
 	RootCmd.AddCommand(daemonCmd)
 }

--- a/command/daemon.go
+++ b/command/daemon.go
@@ -27,7 +27,7 @@ var daemonCmd = &cobra.Command{
 
 func init() {
 	daemonCmd.Flags().IntVarP(&port, "port", "p", 6666, "Local daemon port to listen on")
-	daemonCmd.Flags().StringVarP(&network, "network", "n", "", "Local network interface to listen on ('tcp', 'tcp4', 'tcp6')")
+	daemonCmd.Flags().StringVarP(&network, "network", "n", "tcp", "Local network interface to listen on ('tcp', 'tcp4', 'tcp6')")
 	daemonCmd.Flags().StringVarP(&address, "address", "a", "", "Local network address to listen on (e.g. '', '127.0.0.1', '[::1]' etc.)")
 	RootCmd.AddCommand(daemonCmd)
 }

--- a/command/daemon_test.go
+++ b/command/daemon_test.go
@@ -30,6 +30,7 @@ func TestDaemonCommand(t *testing.T) {
 	args := []string{"daemon"}
 	p, _ := utils.GetFreePort()
 	port = p
+	network = "tcp"
 	go daemonCmd.Run(nil, args)
 
 	waitForPortInTest(port, t)

--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -46,8 +46,8 @@ func NewDaemon(MockServiceManager Service, verificationServiceManager Service) *
 }
 
 // StartDaemon starts the daemon RPC server.
-func (d Daemon) StartDaemon(port int) {
-	log.Println("[INFO] daemon - starting daemon on port", port)
+func (d Daemon) StartDaemon(port int, network string, address string) {
+	log.Println("[INFO] daemon - starting daemon on network:", network, "address:", address, "port:", port)
 
 	serv := rpc.NewServer()
 	serv.Register(d)
@@ -62,7 +62,7 @@ func (d Daemon) StartDaemon(port int) {
 	// Workaround for multiple RPC ServeMux's
 	http.DefaultServeMux = oldMux
 
-	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	l, err := net.Listen(network, fmt.Sprintf("%s:%d", address, port))
 	if err != nil {
 		panic(err)
 	}

--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -97,7 +97,7 @@ func (d Daemon) Shutdown() {
 // StartServer starts a mock server and returns a pointer to atypes.MockServer
 // struct.
 func (d Daemon) StartServer(request types.MockServer, reply *types.MockServer) error {
-	log.Println("[DEBUG] daemon - starting mock server")
+	log.Println("[DEBUG] daemon - starting mock server with args:", request.Args)
 	server := &types.MockServer{}
 	port, svc := d.pactMockSvcManager.NewService(request.Args)
 	server.Port = port

--- a/daemon/daemon_test.go
+++ b/daemon/daemon_test.go
@@ -67,7 +67,7 @@ func TestNewDaemon(t *testing.T) {
 func TestStopDaemon(t *testing.T) {
 	d, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
-	go d.StartDaemon(port)
+	go d.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 	var res string
 	d.StopDaemon("", &res)
@@ -77,7 +77,7 @@ func TestStopDaemon(t *testing.T) {
 func TestShutdownDaemon(t *testing.T) {
 	d, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
-	go d.StartDaemon(port)
+	go d.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 	d.Shutdown()
 }
@@ -129,7 +129,7 @@ func TestStartAndStopDaemon(t *testing.T) {
 	port, _ := utils.GetFreePort()
 	daemon, _ := createMockedDaemon(true)
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 }
 
@@ -339,7 +339,7 @@ func TestRPCClient_List(t *testing.T) {
 	daemon, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 
 	client, err := rpc.DialHTTP("tcp", fmt.Sprintf(":%d", port))
@@ -358,7 +358,7 @@ func TestRPCClient_StartServer(t *testing.T) {
 	daemon, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 
 	client, err := rpc.DialHTTP("tcp", fmt.Sprintf(":%d", port))
@@ -381,7 +381,7 @@ func TestRPCClient_StopServer(t *testing.T) {
 	daemon, manager := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 
 	var cmd *exec.Cmd
@@ -412,7 +412,7 @@ func TestRPCClient_StopDaemon(t *testing.T) {
 	daemon, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 
 	client, err := rpc.DialHTTP("tcp", fmt.Sprintf(":%d", port))
@@ -429,7 +429,7 @@ func TestRPCClient_Verify(t *testing.T) {
 	daemon, _ := createMockedDaemon(true)
 	port, _ := utils.GetFreePort()
 	defer waitForDaemonToShutdown(port, daemon, t)
-	go daemon.StartDaemon(port)
+	go daemon.StartDaemon(port, "tcp", "")
 	connectToDaemon(port, t)
 
 	client, err := rpc.DialHTTP("tcp", fmt.Sprintf(":%d", port))

--- a/daemon/mock_service.go
+++ b/daemon/mock_service.go
@@ -20,7 +20,8 @@ func (m *MockService) NewService(args []string) (int, Service) {
 	log.Println("[DEBUG] starting mock service on port:", port)
 
 	m.Args = []string{
-		fmt.Sprintf("--port %d", port),
+		"--port",
+		fmt.Sprintf("%d", port),
 	}
 	m.Args = append(m.Args, args...)
 

--- a/daemon/mock_service_test.go
+++ b/daemon/mock_service_test.go
@@ -4,7 +4,7 @@ import "testing"
 
 func TestMockService_NewService(t *testing.T) {
 	s := &MockService{}
-	port, svc := s.NewService([]string{"--foo bar"})
+	port, svc := s.NewService([]string{"--foo"})
 
 	if port <= 0 {
 		t.Fatalf("Expected non-zero port but got: %d", port)
@@ -14,7 +14,7 @@ func TestMockService_NewService(t *testing.T) {
 		t.Fatalf("Expected a non-nil object but got nil")
 	}
 
-	if s.Args[1] != "--foo bar" {
-		t.Fatalf("Expected '--foo bar' argument to be passed")
+	if s.Args[2] != "--foo" {
+		t.Fatalf("Expected '--foo' argument to be passed")
 	}
 }

--- a/daemon/service_manager.go
+++ b/daemon/service_manager.go
@@ -69,6 +69,8 @@ func (s *ServiceManager) Stop(pid int) (bool, error) {
 	go func() {
 		s.commandCompleteChan <- cmd
 	}()
+
+	// Wait for error, kill if it takes too long
 	var err error
 	done := make(chan error, 1)
 	go func() {
@@ -78,20 +80,18 @@ func (s *ServiceManager) Stop(pid int) (bool, error) {
 	select {
 	case <-time.After(3 * time.Second):
 		if err = cmd.Process.Kill(); err != nil {
-			log.Println("failed to kill: ", err)
+			log.Println("[ERROR] timeout reached, killing pid", pid)
+
 			return false, err
 		}
-		log.Println("process killed as timeout reached")
 	case err = <-done:
 		if err != nil {
-			log.Printf("process done with error = %v", err)
-
-		} else {
-			log.Print("process done gracefully without error")
+			log.Println("[ERROR] error waiting for process to complete", err)
+			return false, err
 		}
 	}
 
-	return cmd.ProcessState.Success(), nil
+	return true, nil
 }
 
 // List all Service PIDs.

--- a/daemon/service_manager.go
+++ b/daemon/service_manager.go
@@ -2,11 +2,11 @@ package daemon
 
 import (
 	"bufio"
-	"fmt"
 	"io"
 	"log"
 	"os"
 	"os/exec"
+	"time"
 )
 
 // ServiceManager is the default implementation of the Service interface.
@@ -60,6 +60,7 @@ func (s *ServiceManager) removeServiceMonitor() {
 }
 
 // Stop a Service and returns the exit status.
+// Stop a Service and returns the exit status.
 func (s *ServiceManager) Stop(pid int) (bool, error) {
 	log.Println("[DEBUG] stopping service with pid", pid)
 	cmd := s.processes[pid]
@@ -68,11 +69,26 @@ func (s *ServiceManager) Stop(pid int) (bool, error) {
 	go func() {
 		s.commandCompleteChan <- cmd
 	}()
+	var err error
+	done := make(chan error, 1)
+	go func() {
+		done <- cmd.Wait()
+	}()
 
-	err := cmd.Wait()
-	if err != nil {
-		fmt.Fprintln(os.Stderr, "Error waiting for Cmd", err)
-		return false, err
+	select {
+	case <-time.After(3 * time.Second):
+		if err = cmd.Process.Kill(); err != nil {
+			log.Println("failed to kill: ", err)
+			return false, err
+		}
+		log.Println("process killed as timeout reached")
+	case err = <-done:
+		if err != nil {
+			log.Printf("process done with error = %v", err)
+
+		} else {
+			log.Print("process done gracefully without error")
+		}
 	}
 
 	return cmd.ProcessState.Success(), nil

--- a/dsl/client.go
+++ b/dsl/client.go
@@ -31,6 +31,12 @@ type Client interface {
 type PactClient struct {
 	// Port the daemon is running on.
 	Port int
+
+	// Network Daemon is listening on
+	Network string
+
+	// Address the Daemon is listening on
+	Address string
 }
 
 // Get a port given a URL
@@ -52,19 +58,19 @@ func getPort(rawURL string) int {
 	return -1
 }
 
-func getHTTPClient(port int) (*rpc.Client, error) {
+func getHTTPClient(port int, network string, address string) (*rpc.Client, error) {
 	log.Println("[DEBUG] creating an HTTP client")
-	err := waitForPort(port, fmt.Sprintf(`Timed out waiting for Daemon on port %d - are you
+	err := waitForPort(port, network, address, fmt.Sprintf(`Timed out waiting for Daemon on port %d - are you
 		sure it's running?`, port))
 	if err != nil {
 		return nil, err
 	}
-	return rpc.DialHTTP("tcp", fmt.Sprintf(":%d", port))
+	return rpc.DialHTTP(network, fmt.Sprintf("%s:%d", address, port))
 }
 
 // Use this to wait for a daemon to be running prior
 // to running tests.
-var waitForPort = func(port int, message string) error {
+var waitForPort = func(port int, network string, address string, message string) error {
 	log.Println("[DEBUG] waiting for port", port, "to become available")
 	timeout := time.After(timeoutDuration)
 
@@ -74,7 +80,7 @@ var waitForPort = func(port int, message string) error {
 			log.Printf("[ERROR] Expected server to start < %s. %s", timeoutDuration, message)
 			return fmt.Errorf("Expected server to start < %s. %s", timeoutDuration, message)
 		case <-time.After(50 * time.Millisecond):
-			_, err := net.Dial("tcp", fmt.Sprintf(":%d", port))
+			_, err := net.Dial(network, fmt.Sprintf("%s:%d", address, port))
 			if err == nil {
 				return nil
 			}
@@ -86,7 +92,7 @@ var waitForPort = func(port int, message string) error {
 func (p *PactClient) StartServer(args []string) *types.MockServer {
 	log.Println("[DEBUG] client: starting a server")
 	var res types.MockServer
-	client, err := getHTTPClient(p.Port)
+	client, err := getHTTPClient(p.Port, p.getNetworkInterface(), p.Address)
 	if err == nil {
 		err = client.Call(commandStartServer, types.MockServer{Args: args}, &res)
 		if err != nil {
@@ -95,7 +101,7 @@ func (p *PactClient) StartServer(args []string) *types.MockServer {
 	}
 
 	if err == nil {
-		waitForPort(res.Port, fmt.Sprintf(`Timed out waiting for Mock Server to
+		waitForPort(res.Port, p.getNetworkInterface(), p.Address, fmt.Sprintf(`Timed out waiting for Mock Server to
 			start on port %d - are you sure it's running?`, res.Port))
 	}
 
@@ -108,11 +114,11 @@ func (p *PactClient) VerifyProvider(request types.VerifyRequest) (string, error)
 
 	port := getPort(request.ProviderBaseURL)
 
-	waitForPort(port, fmt.Sprintf(`Timed out waiting for Provider API to start
+	waitForPort(port, p.getNetworkInterface(), p.Address, fmt.Sprintf(`Timed out waiting for Provider API to start
 		 on port %d - are you sure it's running?`, port))
 
 	var res types.CommandResponse
-	client, err := getHTTPClient(p.Port)
+	client, err := getHTTPClient(p.Port, p.getNetworkInterface(), p.Address)
 	if err == nil {
 		err = client.Call(commandVerifyProvider, request, &res)
 		if err != nil {
@@ -131,7 +137,7 @@ func (p *PactClient) VerifyProvider(request types.VerifyRequest) (string, error)
 func (p *PactClient) ListServers() *types.PactListResponse {
 	log.Println("[DEBUG] client: listing servers")
 	var res types.PactListResponse
-	client, err := getHTTPClient(p.Port)
+	client, err := getHTTPClient(p.Port, p.getNetworkInterface(), p.Address)
 	if err == nil {
 		err = client.Call(commandListServers, types.MockServer{}, &res)
 		if err != nil {
@@ -145,7 +151,7 @@ func (p *PactClient) ListServers() *types.PactListResponse {
 func (p *PactClient) StopServer(server *types.MockServer) *types.MockServer {
 	log.Println("[DEBUG] client: stop server")
 	var res types.MockServer
-	client, err := getHTTPClient(p.Port)
+	client, err := getHTTPClient(p.Port, p.getNetworkInterface(), p.Address)
 	if err == nil {
 		err = client.Call(commandStopServer, server, &res)
 		if err != nil {
@@ -159,7 +165,7 @@ func (p *PactClient) StopServer(server *types.MockServer) *types.MockServer {
 func (p *PactClient) StopDaemon() error {
 	log.Println("[DEBUG] client: stop daemon")
 	var req, res string
-	client, err := getHTTPClient(p.Port)
+	client, err := getHTTPClient(p.Port, p.getNetworkInterface(), p.Address)
 	if err == nil {
 		err = client.Call(commandStopDaemon, &req, &res)
 		if err != nil {
@@ -167,4 +173,13 @@ func (p *PactClient) StopDaemon() error {
 		}
 	}
 	return err
+}
+
+// getNetworkInterface returns a default interface to communicate to the Daemon
+// if none specified
+func (p *PactClient) getNetworkInterface() string {
+	if p.Network == "" {
+		return "tcp"
+	}
+	return p.Network
 }

--- a/dsl/client_test.go
+++ b/dsl/client_test.go
@@ -109,7 +109,7 @@ func createDaemon(port int, success bool) (*daemon.Daemon, *daemon.ServiceMock) 
 	}()
 
 	d := daemon.NewDaemon(svc, svc)
-	go d.StartDaemon(port)
+	go d.StartDaemon(port, "tcp", "")
 	return d, svc
 }
 

--- a/dsl/pact.go
+++ b/dsl/pact.go
@@ -86,8 +86,8 @@ func (p *Pact) Setup() *Pact {
 	if p.Server == nil {
 		args := []string{
 			fmt.Sprintf("--pact-specification-version %d", p.SpecificationVersion),
-			fmt.Sprintf("--pact-dir %s", p.PactDir),
-			fmt.Sprintf("--log %s/pact.log", p.LogDir),
+			fmt.Sprintf("--pact-dir %s", filepath.FromSlash(p.PactDir)),
+			fmt.Sprintf("--log %s/pact.log", filepath.FromSlash(p.LogDir+"/"+"pact.log")),
 			fmt.Sprintf("--consumer %s", p.Consumer),
 			fmt.Sprintf("--provider %s", p.Provider),
 		}

--- a/dsl/pact.go
+++ b/dsl/pact.go
@@ -85,11 +85,16 @@ func (p *Pact) Setup() *Pact {
 
 	if p.Server == nil {
 		args := []string{
-			fmt.Sprintf("--pact-specification-version %d", p.SpecificationVersion),
-			fmt.Sprintf("--pact-dir %s", filepath.FromSlash(p.PactDir)),
-			fmt.Sprintf("--log %s", filepath.FromSlash(p.LogDir+"/"+"pact.log")),
-			fmt.Sprintf("--consumer %s", p.Consumer),
-			fmt.Sprintf("--provider %s", p.Provider),
+			"--pact-specification-version",
+			fmt.Sprintf("%d", p.SpecificationVersion),
+			"--pact-dir",
+			filepath.FromSlash(p.PactDir),
+			"--log",
+			filepath.FromSlash(p.LogDir + "/" + "pact.log"),
+			"--consumer",
+			p.Consumer,
+			"--provider",
+			p.Provider,
 		}
 		client := &PactClient{Port: p.Port}
 		p.pactClient = client

--- a/dsl/pact.go
+++ b/dsl/pact.go
@@ -51,6 +51,16 @@ type Pact struct {
 	// Specify which version of the Pact Specification should be used (1 or 2).
 	// Defaults to 2.
 	SpecificationVersion int
+
+	// Host is the address of the Daemon, Mock and Verification Service runs on
+	// Examples include 'localhost', '127.0.0.1', '[::1]'
+	// Defaults to 'localhost'
+	Host string
+
+	// Network is the network of the Daemon, Mock and Verification Service
+	// Examples include 'tcp', 'tcp4', 'tcp6'
+	// Defaults to 'tcp'
+	Network string
 }
 
 // AddInteraction creates a new Pact interaction, initialising all
@@ -70,6 +80,14 @@ func (p *Pact) Setup() *Pact {
 	p.setupLogging()
 	log.Printf("[DEBUG] pact setup")
 	dir, _ := os.Getwd()
+
+	if p.Network == "" {
+		p.Network = "tcp"
+	}
+
+	if p.Host == "" {
+		p.Host = "localhost"
+	}
 
 	if p.LogDir == "" {
 		p.LogDir = fmt.Sprintf(filepath.Join(dir, "logs"))
@@ -96,7 +114,7 @@ func (p *Pact) Setup() *Pact {
 			"--provider",
 			p.Provider,
 		}
-		client := &PactClient{Port: p.Port}
+		client := &PactClient{Port: p.Port, Network: p.Network, Address: p.Host}
 		p.pactClient = client
 		p.Server = client.StartServer(args)
 	}
@@ -136,7 +154,7 @@ func (p *Pact) Verify(integrationTest func() error) error {
 	p.Setup()
 	log.Printf("[DEBUG] pact verify")
 	mockServer := &MockService{
-		BaseURL:  fmt.Sprintf("http://localhost:%d", p.Server.Port),
+		BaseURL:  fmt.Sprintf("http://%s:%d", p.Host, p.Server.Port),
 		Consumer: p.Consumer,
 		Provider: p.Provider,
 	}
@@ -170,7 +188,7 @@ func (p *Pact) WritePact() error {
 	p.Setup()
 	log.Printf("[DEBUG] pact write Pact file")
 	mockServer := MockService{
-		BaseURL:  fmt.Sprintf("http://localhost:%d", p.Server.Port),
+		BaseURL:  fmt.Sprintf("http://%s:%d", p.Host, p.Server.Port),
 		Consumer: p.Consumer,
 		Provider: p.Provider,
 	}

--- a/dsl/pact.go
+++ b/dsl/pact.go
@@ -87,7 +87,7 @@ func (p *Pact) Setup() *Pact {
 		args := []string{
 			fmt.Sprintf("--pact-specification-version %d", p.SpecificationVersion),
 			fmt.Sprintf("--pact-dir %s", filepath.FromSlash(p.PactDir)),
-			fmt.Sprintf("--log %s/pact.log", filepath.FromSlash(p.LogDir+"/"+"pact.log")),
+			fmt.Sprintf("--log %s", filepath.FromSlash(p.LogDir+"/"+"pact.log")),
 			fmt.Sprintf("--consumer %s", p.Consumer),
 			fmt.Sprintf("--provider %s", p.Provider),
 		}

--- a/dsl/pact_test.go
+++ b/dsl/pact_test.go
@@ -181,7 +181,7 @@ func TestPact_Setup(t *testing.T) {
 func TestPact_Teardown(t *testing.T) {
 	old := waitForPort
 	defer func() { waitForPort = old }()
-	waitForPort = func(int, string) error {
+	waitForPort = func(int, string, string, string) error {
 		return nil
 	}
 	port, _ := utils.GetFreePort()
@@ -199,7 +199,7 @@ func TestPact_Teardown(t *testing.T) {
 func TestPact_VerifyProvider(t *testing.T) {
 	old := waitForPort
 	defer func() { waitForPort = old }()
-	waitForPort = func(int, string) error {
+	waitForPort = func(int, string, string, string) error {
 		return nil
 	}
 	port, _ := utils.GetFreePort()
@@ -221,7 +221,7 @@ func TestPact_VerifyProviderBroker(t *testing.T) {
 	brokerPort := setupMockBroker(false)
 	old := waitForPort
 	defer func() { waitForPort = old }()
-	waitForPort = func(int, string) error {
+	waitForPort = func(int, string, string, string) error {
 		return nil
 	}
 	port, _ := utils.GetFreePort()
@@ -243,7 +243,7 @@ func TestPact_VerifyProviderBrokerNoConsumers(t *testing.T) {
 	brokerPort := setupMockBroker(false)
 	old := waitForPort
 	defer func() { waitForPort = old }()
-	waitForPort = func(int, string) error {
+	waitForPort = func(int, string, string, string) error {
 		return nil
 	}
 	port, _ := utils.GetFreePort()
@@ -264,7 +264,7 @@ func TestPact_VerifyProviderBrokerNoConsumers(t *testing.T) {
 func TestPact_VerifyProviderFail(t *testing.T) {
 	old := waitForPort
 	defer func() { waitForPort = old }()
-	waitForPort = func(int, string) error {
+	waitForPort = func(int, string, string, string) error {
 		return nil
 	}
 	port, _ := utils.GetFreePort()

--- a/examples/consumer.go
+++ b/examples/consumer.go
@@ -21,6 +21,7 @@ func main() {
 		Port:     6666, // Ensure this port matches the daemon port!
 		Consumer: "MyConsumer",
 		Provider: "MyProvider",
+		Host:     "localhost",
 	}
 	defer pact.Teardown()
 

--- a/types/verify_request.go
+++ b/types/verify_request.go
@@ -42,31 +42,37 @@ type VerifyRequest struct {
 func (v *VerifyRequest) Validate() error {
 	v.Args = []string{}
 	if v.ProviderBaseURL != "" {
-		v.Args = append(v.Args, fmt.Sprintf("--provider-base-url %s", v.ProviderBaseURL))
+		v.Args = append(v.Args, "--provider-base-url")
+		v.Args = append(v.Args, v.ProviderBaseURL)
 	} else {
 		return fmt.Errorf("ProviderBaseURL is mandatory.")
 	}
 
 	if len(v.PactURLs) != 0 {
-		v.Args = append(v.Args, fmt.Sprintf("--pact-urls %s", strings.Join(v.PactURLs[:], ",")))
+		v.Args = append(v.Args, "--pact-urls")
+		v.Args = append(v.Args, strings.Join(v.PactURLs[:], ","))
 	} else {
 		return fmt.Errorf("PactURLs is mandatory.")
 	}
 
 	if v.ProviderStatesSetupURL != "" {
-		v.Args = append(v.Args, fmt.Sprintf("--provider-states-setup-url %s", v.ProviderStatesSetupURL))
+		v.Args = append(v.Args, "--provider-states-setup-url")
+		v.Args = append(v.Args, v.ProviderStatesSetupURL)
 	}
 
 	if v.ProviderStatesURL != "" {
-		v.Args = append(v.Args, fmt.Sprintf("--provider-states-url %s", v.ProviderStatesURL))
+		v.Args = append(v.Args, "--provider-states-url")
+		v.Args = append(v.Args, v.ProviderStatesURL)
 	}
 
 	if v.BrokerUsername != "" {
-		v.Args = append(v.Args, fmt.Sprintf("--broker-username %s", v.BrokerUsername))
+		v.Args = append(v.Args, "--broker-username")
+		v.Args = append(v.Args, v.BrokerUsername)
 	}
 
 	if v.BrokerPassword != "" {
-		v.Args = append(v.Args, fmt.Sprintf("--broker-password %s", v.BrokerPassword))
+		v.Args = append(v.Args, "--broker-password")
+		v.Args = append(v.Args, v.BrokerPassword)
 	}
 	return nil
 }

